pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java

```diff
@@ -164,10 +176,58 @@ public Set<Integer> getPartitions() {
     return _partitions;
   }
 
-  @Nullable
   @Override
-  public Map<IndexType<?, ?, ?>, Long> getIndexSizeMap() {
-    return _indexSizeMap;
+  public long getIndexSizeFor(IndexType type) {
+    short indexId = IndexService.getInstance().getNumericId(type);
+    for (int i = 0, n = getIndexTypeSizesCount(); i < n; i++) {
+      if (indexId == getIndexType(i)) {
+        return getIndexSize(i);
+      }
+    }
+
+    return INDEX_NOT_FOUND;
+  }
+
+  // size should be non-negative 48-bit value
+  public void addIndexSize(short indexType, long size) {
+    if (size < 0 || size > SIZE_MASK) {
+      throw new IllegalArgumentException(
+          "Index size should be a non-negative integer value between 0 and " + SIZE_MASK);
+    }
+    long typeAndSize = ((long) indexType) << 48 | (size & SIZE_MASK);
+    _indexTypeSizeList.add(typeAndSize);
+  }
+
+  @Override
+  public short getIndexType(int i) {
+    return (short) ((_indexTypeSizeList.getLong(i) >> 48) & 0xffff);
+  }
+
+  @Override
+  public long getIndexSize(int i) {
+    return (_indexTypeSizeList.getLong(i) & SIZE_MASK);
+  }
+
+  @JsonIgnore
+  @Override
+  public int getIndexTypeSizesCount() {
+    return _indexTypeSizeList.size();
+  }
+
+  // This method is only meant to retain compatibility of deserialization for
+  // /tables/{tableName}/segments/{segmentName}/metadata endpoint .
+  @SuppressWarnings("unused")
+  public JsonNode getIndexSizeMap() {
```

Comment:
Are we ever serialize/deserialize column metadata as json? It is not designed to do so

Reply:
Yes, as mentioned in the comment, index size map is serialized and passed as part of rest endpoint result.